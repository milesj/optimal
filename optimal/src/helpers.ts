import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
