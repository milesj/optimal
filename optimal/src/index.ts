/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import * as schemas from './schemas';

export type Schemas = typeof schemas;

export { schemas };
export * from './createPredicate';
export * from './createSchema';
export * from './criteria';
export * from './optimal';
export * from './OptimalError';
export * from './schemas/array';
export * from './schemas/blueprint';
export * from './schemas/bool';
export * from './schemas/custom';
export * from './schemas/date';
export * from './schemas/func';
export * from './schemas/instance';
export * from './schemas/lazy';
export * from './schemas/number';
export * from './schemas/object';
export * from './schemas/regex';
export * from './schemas/schema';
export * from './schemas/shape';
export * from './schemas/string';
export * from './schemas/tuple';
export * from './schemas/union';
export * from './types';
export * from './ValidationError';
