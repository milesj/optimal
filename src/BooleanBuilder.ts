import Builder from './Builder';
import { DefaultValue } from './types';

export default class BooleanBuilder<T extends boolean = boolean> extends Builder<T> {
  constructor(defaultValue?: DefaultValue<T>) {
    super('boolean', defaultValue || (false as T));
  }

  onlyFalse(): BooleanBuilder<false> {
    this.defaultValue = false as T;

    if (__DEV__) {
      this.addCheck((path, value) => {
        this.invariant(value === false, 'May only be `false`.', path);
      });
    }

    return (this as unknown) as BooleanBuilder<false>;
  }

  onlyTrue(): BooleanBuilder<true> {
    this.defaultValue = true as T;

    if (__DEV__) {
      this.addCheck((path, value) => {
        this.invariant(value === true, 'May only be `true`.', path);
      });
    }

    return (this as unknown) as BooleanBuilder<true>;
  }
}

export function bool(defaultValue: DefaultValue<boolean> = false) /* infer */ {
  return new BooleanBuilder<boolean>(defaultValue);
}
