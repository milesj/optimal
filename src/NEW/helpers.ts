import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
