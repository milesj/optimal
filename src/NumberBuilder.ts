import Builder from './Builder';
import { DefaultValue } from './types';

function isNumber(value: unknown): value is number {
  return typeof value === 'number';
}

export default class NumberBuilder<T extends number = number> extends Builder<T> {
  constructor(defaultValue?: DefaultValue<T>) {
    super('number', defaultValue || (0 as T));
  }

  between(min: number, max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(
        isNumber(min) && isNumber(max),
        'Between requires a minimum and maximum number.',
      );

      this.addCheck((path, value) => {
        this.invariant(
          isNumber(value) &&
            (inclusive ? value >= min && value <= max : value > min && value < max),
          `Number must be between ${min} and ${max}${inclusive ? ' inclusive' : ''}.`,
          path,
        );
      });
    }

    return this;
  }

  gt(min: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(min), 'Greater-than requires a minimum number.');

      this.addCheck((path, value) => {
        if (inclusive) {
          this.invariant(
            isNumber(value) && value >= min,
            `Number must be greater than or equal to ${min}.`,
            path,
          );
        } else {
          this.invariant(
            isNumber(value) && value > min,
            `Number must be greater than ${min}.`,
            path,
          );
        }
      });
    }

    return this;
  }

  gte(min: number): this {
    return this.gt(min, true);
  }

  lt(max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(max), 'Less-than requires a maximum number.');

      this.addCheck((path, value) => {
        if (inclusive) {
          this.invariant(
            isNumber(value) && value <= max,
            `Number must be less than or equal to ${max}.`,
            path,
          );
        } else {
          this.invariant(isNumber(value) && value < max, `Number must be less than ${max}.`, path);
        }
      });
    }

    return this;
  }

  lte(max: number): this {
    return this.lt(max, true);
  }

  oneOf<U extends number>(list: U[]): NumberBuilder<U> {
    if (__DEV__) {
      this.invariant(
        Array.isArray(list) && list.length > 0 && list.every(item => isNumber(item)),
        'One of requires a non-empty array of numbers.',
      );

      this.addCheck((path, value) => {
        this.invariant(list.includes(value), `Number must be one of: ${list.join(', ')}`, path);
      });
    }

    return (this as unknown) as NumberBuilder<U>;
  }
}

export function number<T extends number = number>(defaultValue?: DefaultValue<number>) /* infer */ {
  return new NumberBuilder<T>(defaultValue as T);
}
