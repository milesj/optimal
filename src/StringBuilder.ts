import CollectionBuilder from './CollectionBuilder';
import { DefaultValue } from './types';

function isString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export default class StringBuilder<T extends string = string> extends CollectionBuilder<T> {
  constructor(defaultValue?: DefaultValue<T>) {
    super('string', defaultValue || ('' as T));
  }

  camelCase(): this {
    return this.match(/^[a-z][a-zA-Z0-9]+$/u, 'String must be in camel case.');
  }

  contains(token: string, index: number = 0): this {
    if (__DEV__) {
      this.invariant(isString(token), 'Contains requires a non-empty token.');

      this.addCheck((path, value) => {
        if (this.isOptionalDefault(value)) {
          return;
        }

        this.invariant(value.includes(token, index), `String does not include "${token}".`, path);
      });
    }

    return this;
  }

  kebabCase(): this {
    return this.match(/^[a-z][a-z0-9-]+$/u, 'String must be in kebab case.');
  }

  match(pattern: RegExp, message: string = ''): this {
    if (__DEV__) {
      this.invariant(
        pattern instanceof RegExp,
        'Match requires a regular expression to match against.',
      );

      this.addCheck((path, value) => {
        if (this.isOptionalDefault(value)) {
          return;
        }

        this.invariant(
          !!value.match(pattern),
          `${message || 'String does not match.'} (pattern "${pattern.source}")`,
          path,
        );
      });
    }

    return this;
  }

  notEmpty(): this {
    if (__DEV__) {
      this.addCheck((path, value) => {
        this.invariant(isString(value), 'String cannot be empty.', path);
      });
    }

    return this;
  }

  oneOf<U extends string>(list: U[]): StringBuilder<U> {
    if (__DEV__) {
      this.invariant(
        Array.isArray(list) && list.length > 0 && list.every(item => isString(item)),
        'One of requires a non-empty array of strings.',
      );

      this.addCheck((path, value) => {
        this.invariant(list.includes(value), `String must be one of: ${list.join(', ')}`, path);
      });
    }

    return (this as unknown) as StringBuilder<U>;
  }

  pascalCase(): this {
    return this.match(/^[A-Z][a-zA-Z0-9]+$/u, 'String must be in pascal case.');
  }

  snakeCase(): this {
    return this.match(/^[a-z][a-z0-9_]+$/u, 'String must be in snake case.');
  }
}

export function string<T extends string = string>(defaultValue?: DefaultValue<string>) /* infer */ {
  return new StringBuilder<T>(defaultValue as T);
}
