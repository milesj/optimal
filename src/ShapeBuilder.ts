import Builder from './Builder';
import isObject from './isObject';
import optimal from './optimal';
import { Blueprint, OptimalOptions } from './types';

export default class ShapeBuilder<Shape extends object> extends Builder<Shape> {
  protected contents: Blueprint<Shape>;

  protected isExact: boolean = false;

  constructor(contents: Blueprint<Shape>) {
    // @ts-ignore
    super('shape', {});

    if (__DEV__) {
      this.invariant(
        isObject(contents) &&
          Object.keys(contents).length > 0 &&
          Object.keys(contents).every(
            key => contents[key as keyof typeof contents] instanceof Builder,
          ),
        'A non-empty object of properties to blueprints are required for a shape.',
      );
    }

    this.contents = contents;
  }

  exact(): this {
    this.isExact = true;

    return this;
  }

  runChecks(
    path: string,
    initialValue: Shape | undefined,
    struct: object,
    options: OptimalOptions = {},
  ) {
    const object = initialValue || this.defaultValue || {};

    if (__DEV__) {
      this.invariant(isObject(object), 'Value passed to shape must be an object.', path);
    }

    return optimal(object, this.contents, {
      ...options,
      prefix: path,
      unknown: !this.isExact,
    });
  }
}

export function shape<P extends object>(contents: Blueprint<P>) /* infer */ {
  return new ShapeBuilder<P>(contents);
}
