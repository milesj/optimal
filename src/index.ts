/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal from './optimal';
import Builder, { custom, func } from './Builder';
import ArrayBuilder, { array } from './ArrayBuilder';
import BooleanBuilder, { bool } from './BooleanBuilder';
import InstanceBuilder, { builder, instance, date, regex } from './InstanceBuilder';
import NumberBuilder, { number } from './NumberBuilder';
import ObjectBuilder, { object, blueprint } from './ObjectBuilder';
import ShapeBuilder, { shape } from './ShapeBuilder';
import StringBuilder, { string } from './StringBuilder';
import TupleBuilder, { tuple } from './TupleBuilder';
import UnionBuilder, { union } from './UnionBuilder';

export const predicates = {
  array,
  blueprint,
  bool,
  builder,
  custom,
  date,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  array,
  blueprint,
  bool,
  builder,
  custom,
  date,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  Builder,
  ArrayBuilder,
  BooleanBuilder,
  InstanceBuilder,
  NumberBuilder,
  ObjectBuilder,
  ShapeBuilder,
  StringBuilder,
  TupleBuilder,
  UnionBuilder,
};

export type Predicates = typeof predicates;

export * from './types';

export default optimal;
