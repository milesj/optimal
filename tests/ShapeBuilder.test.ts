/* eslint-disable @typescript-eslint/no-explicit-any */

import ShapeBuilder, { shape } from '../src/ShapeBuilder';
import { bool } from '../src/BooleanBuilder';
import { number } from '../src/NumberBuilder';
import { string } from '../src/StringBuilder';
import { runChecks, runInProd } from './helpers';

describe('ShapeBuilder', () => {
  let builder: ShapeBuilder<{
    foo: string;
    bar: number;
    baz: boolean;
  }>;

  beforeEach(() => {
    builder = shape({
      foo: string(),
      bar: number(),
      baz: bool(),
    });
  });

  it('errors if a non-object is not passed', () => {
    expect(() => {
      // @ts-ignore
      shape('foo');
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an empty object is passed', () => {
    expect(() => {
      shape({});
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an object with non-builders is passed', () => {
    expect(() => {
      // @ts-ignore
      shape({ foo: 123 });
    }).toThrowErrorMatchingSnapshot();
  });

  it('doesnt error if a builder object is passed', () => {
    expect(() => {
      shape({
        foo: string(),
      });
    }).not.toThrow();
  });

  describe('run()', () => {
    it('errors if a non-object is passed', () => {
      expect(() => {
        runChecks(
          builder,
          // @ts-ignore Allow invalid type
          'foo',
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors for unknown fields when using exact', () => {
      expect(() => {
        builder.exact();

        runChecks(
          builder,
          // @ts-ignore Allow invalid fields
          { qux: 123, oof: 'abc' },
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error for unknown fields if unknown is true', () => {
      expect(() => {
        runChecks(
          builder,
          // @ts-ignore Allow invalid fields
          { qux: 123, oof: 'abc' },
        );
      }).not.toThrow();
    });

    it('checks each item in the object', () => {
      expect(() => {
        runChecks(
          builder,
          // @ts-ignore Allow invalid type
          {
            foo: 'foo',
            bar: 'bar',
            baz: true,
          },
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if an object item is invalid; persists path with index', () => {
      expect(() => {
        runChecks(
          builder,
          // @ts-ignore Allow invalid type
          {
            foo: 123,
          },
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('supports shapes of shapes', () => {
      const nestedBuilder = shape({
        foo: shape({
          a: number(),
          b: number(),
          c: string(),
        }),
      });

      const data = {
        foo: {
          a: 123,
          b: 456,
        },
      };

      expect(runChecks(nestedBuilder, data as any)).toEqual({
        foo: {
          ...data.foo,
          c: '',
        },
      });
    });

    it('supports nested required', () => {
      const nestedBuilder = shape({
        foo: string(),
        bar: bool().required(),
      });

      expect(() => {
        runChecks(nestedBuilder, {
          foo: 'abc',
        } as any);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors correctly for shapes in shapes', () => {
      const nestedBuilder = shape({
        foo: shape({
          a: number(),
          b: number(),
          c: string(),
        }),
      });

      expect(() => {
        runChecks(
          nestedBuilder,
          // @ts-ignore Allow invalid type
          {
            foo: {
              a: 123,
              b: 456,
              c: 789,
            },
          },
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('should be the object with defaults', () => {
      expect(runChecks(builder, {} as any)).toEqual({
        bar: 0,
        baz: false,
        foo: '',
      });
    });

    describe('production', () => {
      it(
        'returns default shape if value is object',
        runInProd(() => {
          expect(runChecks(builder, {})).toEqual({ foo: '', bar: 0, baz: false });
        }),
      );

      it(
        'returns default shape if value is undefined',
        runInProd(() => {
          const def = { foo: 'foo', bar: 123, baz: true };
          builder.defaultValue = def;

          expect(runChecks(builder)).toEqual(def);
        }),
      );

      it(
        'bypasses checks and returns value',
        runInProd(() => {
          expect(
            runChecks(
              builder,
              // @ts-ignore Test invalid type
              { foo: 123 },
            ),
          ).toEqual({ foo: '123', bar: 0, baz: false });
        }),
      );
    });
  });

  describe('typeAlias()', () => {
    it('returns the type name', () => {
      expect(
        shape({
          a: number(),
          b: number(),
          c: string(),
        }).typeAlias(),
      ).toBe('shape');
    });
  });
});
