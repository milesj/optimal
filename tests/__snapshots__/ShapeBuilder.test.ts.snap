// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shape() errors if a non-object is not passed 1`] = `"A non-empty object of properties to blueprints are required for a shape."`;

exports[`shape() errors if an empty object is passed 1`] = `"A non-empty object of properties to blueprints are required for a shape."`;

exports[`shape() errors if an object with non-builders is passed 1`] = `"A non-empty object of properties to blueprints are required for a shape."`;

exports[`shape() runChecks() checks each item in the object 1`] = `"Invalid field \\"key.bar\\". Must be a number."`;

exports[`shape() runChecks() errors correctly for shapes in shapes 1`] = `"Invalid field \\"key.foo.c\\". Must be a string."`;

exports[`shape() runChecks() errors for unknown fields when using exact 1`] = `"Unknown \\"key\\" fields: qux, oof."`;

exports[`shape() runChecks() errors if a non-object is passed 1`] = `"Invalid field \\"key\\". Value passed to shape must be an object."`;

exports[`shape() runChecks() errors if an object item is invalid; persists path with index 1`] = `"Invalid field \\"key.foo\\". Must be a string."`;

exports[`shape() runChecks() supports nested required 1`] = `"Invalid field \\"key.bar\\". Field is required and must be defined."`;
